src/profile-logic/perf-script-profile-format.js

```diff
@@ -0,0 +1,219 @@
+        stackTable.data.push([frame, prefix]);
+        stackMap.set(key, stack);
+      }
+      return stack;
+    }
+
+    const frameMap = new Map();
+    function getOrCreateFrame(frameString) {
+      let frame = frameMap.get(frameString);
+      if (frame === undefined) {
+        frame = frameTable.data.length;
+        const stringIndex = stringTable.length;
+        stringTable.push(frameString);
+        frameTable.data.push([stringIndex]);
+        frameMap.set(frameString, frame);
+      }
+      return frame;
+    }
+
+    function addSample(threadName, stackArray, time) {
+      // often we create a thread which inherits the name of the parent, and
+      // set the thread's name slightly later.  Avoid having the first
+      // sample's name stick.
+      if (name !== threadName) {
+        name = threadName;
+      }
+      const stack = stackArray.reduce((prefix, stackFrame) => {
+        const frame = getOrCreateFrame(stackFrame);
+        return getOrCreateStack(frame, prefix);
+      }, null);
+      samples.data.push([stack, time]);
+    }
+
+    return {
+      addSample,
+      finish: () => {
+        return {
+          tid,
+          pid,
+          name,
+          markers,
+          samples,
+          frameTable,
+          stackTable,
+          stringTable,
+        };
+      },
+    };
+  }
+
+  const threadMap = new Map();
+
+  function _addThreadSample(pid, tid, threadName, timeStamp, stack) {
+    // Right now this assumes that you can't have two identical tids in
+    // different pids, which is true in linux at least.
+    let thread = threadMap.get(tid);
+    if (thread === undefined) {
+      thread = _createThread(threadName, pid, tid);
+      threadMap.set(tid, thread);
+    }
+    thread.addSample(threadName, stack, timeStamp);
+  }
+
+  // Parse the format. The two regular expressions and some of the comments
+  // below were taken from Brendan Gregg's stackcollapse-perf.pl.
+  const lines = profile.split('\n');
+
+  let lineIndex = 0;
+  let startTime = 0;
+  while (lineIndex < lines.length) {
+    const sampleStartLine = lines[lineIndex++];
+    // default "perf script" output has TID but not PID
+    // eg, "java 25607 4794564.109216: cycles:"
+    // eg, "java 12688 [002] 6544038.708352: cpu-clock:"
+    // eg, "V8 WorkerThread 25607 4794564.109216: cycles:"
+    // eg, "java 24636/25607 [000] 4794564.109216: cycles:"
+    // eg, "java 12688/12764 6544038.708352: cpu-clock:"
+    // eg, "V8 WorkerThread 24636/25607 [000] 94564.109216: cycles:"
+    // other combinations possible
+    // pattern: thread-name-with-optional-spaces tid time: NNN cycles:XXXX:
+    // alternate pattern: thread-name-with-optional-spaces pid/tid time: NNN cycles:XXXX:
+    // eg, "java 25607/25608 4794564.109216: 33 cycles:uppp"
+    //   (generate with "perf script -F +pid")
+    const sampleStartMatch = /^(\S.*?)(?=(?:\s+(?:(?:\d+)\/*(?:\d+)*\s)))\s+(\d+)\/*(\d+)*\s+([\d.]+)/.exec(
```

Comment:
Initially I was going to suggest documenting and testing this regex, as it's not very maintainable as it is. However, I'd be fine if we had a permalink to a specific revision that this regex was taken from, that way we could upstream any changes needed.

Reply:
Regexes are horrible.... Closest thing to APL in modern coding.  I will document it, since I had to add a bunch to it to handle spaces in threadnames, and to handle the pid.